server/src/main/java/org/opensearch/common/metadata/MetadataParser.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.metadata;
+
+import java.io.IOException;
+
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.store.BufferedChecksumIndexInput;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+
+/**
+ * Manages Metadata by adding version, codec in header and checksum in footer.
+ * @param <T> The type of metadata to be read/written
+ */
+public class MetadataParser<T> {
+    // This can be updated to hold a parserFactory and get relevant parsers based on the metadata versions
+    private final MetadataContentParser<T> parser;
+    private final int currentVersion;
+    private final String codec;
+
+    /**
+     * @param parser parser to read/write metadata from T
+     * @param currentVersion latest supported version of the metadata
+     * @param codec: metadata codec
+     */
+    public MetadataParser(MetadataContentParser<T> parser, int currentVersion, String codec) {
+        this.parser = parser;
+        this.currentVersion = currentVersion;
+        this.codec = codec;
+    }
+
+    /**
+     * Reads metadata content from {@code indexInput} and parses the read content to {@link T}.
+     * Before reading actual content, verifies the header with relevant codec and version.
+     * After reading the actual content, verifies the checksum as well
+     * @param indexInput metadata file input stream
+     * @return metadata content parsed into {@link T}
+     */
+    public T readMetadata(IndexInput indexInput) throws IOException {
+        ChecksumIndexInput checksumIndexInput = new BufferedChecksumIndexInput(indexInput);
+        checkHeader(checksumIndexInput);
+        T metadata = this.parser.readContent(checksumIndexInput);
+        checkFooter(checksumIndexInput);
+        return metadata;
+    }
+
+    /**
+     * Writes metadata to file output stream {@code indexOutput}
+     * @param indexOutput file output stream which will store metadata content
+     * @param metadata metadata content.
+     */
+    public void writeMetadata(IndexOutput indexOutput, T metadata) throws IOException {
+        this.writeHeader(indexOutput);
+        this.parser.writeContent(indexOutput, metadata);
+        this.writeFooter(indexOutput);
+    }
```

Comment:
Where are we writing the version of the current metadata being written

Reply:
In `this.writeHeader`, the `this.currentVersion` is written to header as metadata version. ln: 89

we get this version when metadata file is read. ln:49.

Reply:
Please add an explicit TODO in the code line:49 is confusing is since it isn't consuming the version, nor do contracts `readContent ` specify they will use a reader which is version aware

Reply:
Sure @Bukhtawar. We'll make it more explicit to indicate this.